Simple check docu (#113)
hpi-bp1819-naumann/shift-detector: docs/source/checks/simple.rst

.. _simple:

> Simple Check

Reviewer: ```suggestion
Simple
```

stay consistent

Author: yup

> ====================

Reviewer: ```suggestion
======
```

Author: not really that important I would say, but okay

Description
-----------

> The SimpleCheck gives you basic information about your datasets and detects the most obvious of datashifts.

Reviewer: Use references here
```suggestion
:ref:`simple` gives you basic information about your datasets and detects the most obvious of datashifts.
```

Author: Also, I don't really get why it would make sense to reference the file you are in. But for consistency reasons .. ok

Author: Also, I don't really get why it would make sense to reference the file you are in. But for consistency reasons .. ok

Some shifts result in the change of basic metrics, that are commonly used for the analysis and exploration of datasets.
Those metrics and statistics are calculated on both datasets and compared afterwards. HIER NOCH AMAZON DEEQU ERWÄHNEN,

> WIRD NACH UMBENENNUNG DES CHECKS GEMACHT

Reviewer: Create an Issue for that and delete the comment from the documentation

Author: Good idea. [#125](https://github.com/hpi-bp1819-naumann/shift-detector/issues/125)

> The basic check should be used in most analyses because its metrics are scientifically grounded and commmonly used.

Reviewer: I don't like this sentence - I don't think that 
```suggestion
:ref:`simple` should be used in most analyses because its metrics are scientifically grounded and commmonly used.
```

Author: Alright, I weakend it a little bit. 
"The basic check should be used in most analyses, its metrics are commonly used to describe datasets."

> The basic check should be used in most analyses because its metrics are scientifically grounded and commmonly used.

Reviewer: I don't really like this sentence at all. The other checks are scientifically grounded too. The argumentation that you should use it, because it's commonly used isn't good too...

Author: doubled to the comment above? If not Im not sure to which line you are referring to in the comment above. 

Thats not true. Things that are commonly used in science are a good indicator for a solid metric. That has nothing to do with the check being so great, it goes more in the direction that metrics like that are standard metrics to describe data in datascience. Which is a good indicator of the metric being useful.

It works directly on categorical and numerical data and, with additional Precalculations, also on text data.

Example
-------

::

>     from shift_detector.detector import Detector

Reviewer: ```suggestion
    from Morpheus.detector import Detector
```

Author: ok

>     from shift_detector.checks.simple_check import SimpleCheck

Reviewer: ```suggestion
    from Morpheus.checks.simple_check import SimpleCheck
```

Author: ok

    data_set_1 = 'examples/pokedex1.csv'

>     data_set_2 = 'examples/pokedex2.csv'

Reviewer: Until now, the documentation uses `shoes_first.csv` and `shoes_second.csv`... I think it is a good idea to have an example, the user can reproduce with data provided data, but the documentation should be consistent. So please change this or create an issue for the other checks

Author: created an issue #126

    detector = Detector(
        data_set_1,
        data_set_2,
        delimiter=','
    )

    detector.run(SimpleCheck())
    detector.evaluate()

The code works as follows:

> 1. First, you create a :class:`~shift_detector.detector.Detector` object to tell Morpheus

Reviewer: ```suggestion
1. First, you create a :class:`~Morpheus.detector.Detector` object to tell Morpheus
```

Author: ok

   which data sets you want to compare.
2. Then you start the detector with

>    :meth:`~shift_detector.detector.Detector.run` and the checks you want to run: in this case

Reviewer: ```suggestion
   :meth:`~Morpheus.detector.Detector.run(...)` and the checks you want to run: in this case
```

Author: ok

>    :class:`~shift_detector.checks.simple_check.SimpleCheck`.

Reviewer: ```suggestion
   :class:`~Morpheus.checks.simple_check.SimpleCheck()`.
```

Author: ok

3. Finally, you print the result with

>    :meth:`~shift_detector.detector.Detector.evaluate`

Reviewer: ```suggestion
   :meth:`~Morpheus.detector.Detector.evaluate()`
```

Author: k

Result
++++++

The SimpleCheck produces this report:

::

    Numerical Columns:

    Column 'payment':
    Metric: mean with Diff: -29.4 %, threshold: 20%
    Metric: std with Diff: +80.2 %, threshold: 50%

    (...)

    Categorical Columns

    Column 'payment_option':
    Attribute: 'debit' with Diff: +6.0 %, categorical threshold: 5%
    Attribute: 'cash' with Diff: -6.5 %, categorical threshold: 5%

    (...)

Interpretation
++++++++++++++

The results show column-wise wich shifts appear in your dataset. First the reports of your numerical columns are shown,

> after that the reports of your categorical columns.

Reviewer: What about textual columns? You wrote about them earlier

Author: youre right. see #127

Numerical
~~~~~~~~~

In the example, below the heading 'numerical columns', you see that the column with the name 'payment' is detected

> as shifted. You see below the columnname that two numerical metrics detected a shift, *mean* and *std*

Reviewer: ```suggestion
as shifted. Below the columnname, you can see that two numerical metrics detected a shift, *mean* and *std*
```

just a small fix, that makes me feel better

Author: changed

> In the numerical columns you see that the column with the name 'payment' is detected as shifted. You see below

Reviewer: Where exactly can I see this?

Author: in the example ;) 
fixed it

the columnname that two numerical metrics detected a shift, *mean* and *std* (standard deviation).

*Diff* describes the relative value-distance of the metrics that are calculated on your two datasets. If the Diff
exceeds a metric-specific threshold, the dataset differ that much, that the check calls shift on this column.

In this case, the mean-value of 'payment' could be 1000.0 in your first, but 706.0 in your second dataset. This difference is
results in an difference of -294 and an relative difference of -29.4%. The threshold set here is 20% which is exceeded
by the absolute value of -29.4%. All thresholds are customizable through the API of the

> :class:`~shift_detector.checks.simple_check.SimpleCheck`. All the metrics are listed in the section

Reviewer: ```suggestion
:class:`~Morpheus.checks.simple_check.SimpleCheck`. All the metrics are listed in the section
```

Author: yo

:ref:`simple_check_parameters`.

Categorical
~~~~~~~~~~~

Using the SimpleCheck shifts can also be detected on categorical columns. Here the histograms over the attribute-values are
compared. If the difference between those values exceeds the *categorical_threshold* the check calls shift on this
column.

The example above shows a shift in the categorical column 'payment_option'. There are different
attribute-values in this column, i.e. 'debit and 'cash'. In those, the differences between the dataset are 6% and

> -6.5% which both exceed the threshold of 5%. This indicates that the in dataset2 more people use

Reviewer: ```suggestion
-6.5% which both exceed the threshold of 5%. This indicates that in dataset2 more people use
```

Author: ah, thanks

cash as a payment-option and fewer use debit, the check calls shift on the column.

.. _simple_check_parameters:

Metrics & Parameters

> ----------

Reviewer: line to short

There are 9 different numerical metrics in the SimpleCheck whose differences can indicate a shift. All
default-thresholds can be adjusted.

+-----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| metric_name                       |                                                                           explanation                                                   |
+===================================+=========================================================================================================================================+
| **quartile_1,                     |                                                                                                                                         |

> | median,                           | Those are the .25- .5- and .75-quantiles of the column.                                                                                 |

Reviewer: is quartile an quantile the same? Can we use consistent names?

Author: Here I use the more common term 'quantile' to describe the term 'quartile'. A the quartiles are the quantiles that are dividable by 0.25

| quartile_3**                      |                                                                                                                                         |
+-----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+

> | **value-range**                   | The range of values of the columns, calculated as maximum minus minimum of that column.                                                 |

Reviewer: ```suggestion
| **value-range**                   | The range of values of the columns, calculated as maximum - minimum of that column.                                                 |
```

better readability

Author: found a nicer way, thanks

+-----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+

> | **std**                           | the standard deviation in the column                                                                                                    |

Reviewer: std seems a little bit vacuous. Can we rename it to `stddev`?

Author: its commonly used like that see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.std.html

| **completeness**                  | The ratio of non None-values to the total number of values.                                                                             |
+-----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| **num_distinct**                  | The total number of distinct values in a column.                                                                                        |
+-----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+

All parameters expect float values larger than 0.0. The smallest value of 0.0 means that only the smallest of changes
(bigger than 0%) is going to result in the metric to call shift. For most cases a threshold smaller then 1.0, meaning
100% difference is recommended.

The 9 threshold parameters for numerical metrics are accessable through their name *'metric_name'_threshold*,
the threshold parameter for the categorical columns is called *categorical_threshold*.

Example
::

>     from shift_detector.checks.simple_check import SimpleCheck

Reviewer: ```suggestion
    from Morpheus.checks.simple_check import SimpleCheck
```

Author: k

    sc = new SimpleCheck(median_threshold=.05, std_threshold=.42, categorical_threshold=1.05)

Implementation
--------------

Algorithm
+++++++++

> The SimpleCheck works as follows:

Reviewer: ```suggestion
:ref:`simple` works as follows:
```

Author: ok

> 1.  First, calculating the metrics for all usable columns of the datasets ds1 and ds2. Most metrics are build upon

Reviewer: ```suggestion
1.  First, calculate the metrics for all usable columns of the datasets ds1 and ds2. Most metrics are build upon
```

Author: alrighty

    functions from the python library *pandas*
2.  Then, take the difference between each metric, so *diff_metric = metric(ds1) - metric(ds2)*
3.  Finally, compare those diffs to the predifined or custom thresholds. If the threshold is exceeded, indicate a shift

Notes
+++++

The 0.0- and 1.0-quantiles, so the minima and maxima, are not part of the shift-metrics because they have proven to be
very unresistant to outliers.

References
----------

> To be filled with Deequ Paper...

Reviewer: Please do this...

Author: see #125
